src/openms/include/OpenMS/METADATA/ID/ObservationMatch.h

```diff
@@ -0,0 +1,154 @@
+
+      AdductOpt adduct_opt; ///< optional reference to adduct
+
+      // peak annotations (fragment ion matches), potentially from different
+      // data processing steps:
+      PeakAnnotationSteps peak_annotations;
+
+      explicit ObservationMatch(
+        IdentifiedMolecule identified_molecule_var,
+        ObservationRef observation_ref, Int charge = 0,
+        const boost::optional<AdductRef>& adduct_opt = boost::none,
+        const AppliedProcessingSteps& steps_and_scores = AppliedProcessingSteps(),
+        const PeakAnnotationSteps& peak_annotations = PeakAnnotationSteps()):
+        ScoredProcessingResult(steps_and_scores),
+        identified_molecule_var(identified_molecule_var),
+        observation_ref(observation_ref), charge(charge), adduct_opt(adduct_opt),
+        peak_annotations(peak_annotations)
+      {
+      }
+
+      ObservationMatch(const ObservationMatch&) = default;
+
+      ObservationMatch& merge(const ObservationMatch& other)
+      {
+        ScoredProcessingResult::merge(other);
+        if (charge == 0)
+        {
+          charge = other.charge;
+        }
+        else if (charge != other.charge)
+        {
+          throw Exception::InvalidValue(__FILE__, __LINE__,
+                                        OPENMS_PRETTY_FUNCTION,
+                                        "Trying to overwrite ObservationMatch charge with conflicting value.",
+                                        String(charge));
+        }
+
+        if (!adduct_opt)
+        {
+          adduct_opt = other.adduct_opt;
+        }
+        else if (adduct_opt != other.adduct_opt)
+        {
+          throw Exception::InvalidValue(__FILE__, __LINE__,
+                                        OPENMS_PRETTY_FUNCTION,
+                                        "Trying to overwrite ObservationMatch adduct_opt with conflicting value.",
+                                        (*adduct_opt)->getName());
+        }
+
+        peak_annotations.insert(other.peak_annotations.begin(),
+                                other.peak_annotations.end());
+        return *this;
+      }
+    };
+
+    // all matches for the same observation should be consecutive, so make sure
+    // the observation is used as the first member in the composite key:
+    typedef boost::multi_index_container<
+      ObservationMatch,
+      boost::multi_index::indexed_by<
+        boost::multi_index::ordered_unique<
+          boost::multi_index::composite_key<
+            ObservationMatch,
+            boost::multi_index::member<ObservationMatch, ObservationRef,
+                                       &ObservationMatch::observation_ref>,
+            boost::multi_index::member<
+              ObservationMatch, IdentifiedMolecule,
+              &ObservationMatch::identified_molecule_var>,
+            boost::multi_index::member<ObservationMatch, AdductOpt,
+                                       &ObservationMatch::adduct_opt>>>>
+      > ObservationMatches;
```

Comment:
easy. I could write this any time of the day...
:P

Reply:
That's C++ template magic for you. If you follow the Boost documentation, it's not too hard to use, though.